File: packages/client/src/checkout.rs
use crate::{
	self as tg,
	util::serde::{is_false, is_true, return_true},
};
use futures::{Stream, StreamExt as _, TryStreamExt as _, future};
use std::path::PathBuf;
use tangram_futures::stream::TryExt as _;
use tangram_http::{request::builder::Ext as _, response::Ext as _};

#[derive(Clone, Debug, serde::Deserialize, serde::Serialize)]
pub struct Arg {
	pub artifact: tg::artifact::Id,

	/// Whether to check out the artifact's dependencies.
	#[serde(default = "return_true", skip_serializing_if = "is_true")]
	pub dependencies: bool,

	#[serde(default, skip_serializing_if = "is_false")]
	pub force: bool,

	#[serde(default = "return_true", skip_serializing_if = "is_true")]
	pub lock: bool,

	#[serde(default, skip_serializing_if = "Option::is_none")]
	pub path: Option<PathBuf>,
}

#[derive(Clone, Debug, serde::Deserialize, serde::Serialize)]
pub struct Output {
	pub path: PathBuf,
}

pub async fn checkout<H>(handle: &H, arg: Arg) -> tg::Result<PathBuf>
where
	H: tg::Handle,
{
	let stream = handle.checkout(arg).await?.boxed();
	let output = stream
		.try_last()
		.await?
		.and_then(|event| event.try_unwrap_output().ok())
		.ok_or_else(|| tg::error!("stream ended without output"))?;
	Ok(output.path)
}

impl tg::Client {
	pub async fn checkout(
		&self,
		arg: tg::checkout::Arg,
	) -> tg::Result<
		impl Stream<Item = tg::Result<tg::progress::Event<tg::checkout::Output>>> + Send + 'static,
	> {
		let method = http::Method::POST;
		let uri = "/checkout";
		let request = http::request::Builder::default()
			.method(method)
			.uri(uri)
			.header(http::header::ACCEPT, mime::TEXT_EVENT_STREAM.to_string())
			.json(arg)
			.map_err(|source| tg::error!(!source, "failed to serialize the arg"))?
			.unwrap();
		let response = self.send(request).await?;
		if !response.status().is_success() {
			let error = response.json().await?;
			return Err(error);
		}
		let content_type = response
			.parse_header::<mime::Mime, _>(http::header::CONTENT_TYPE)
			.transpose()?;
		if !matches!(
			content_type
				.as_ref()
				.map(|content_type| (content_type.type_(), content_type.subtype())),
			Some((mime::TEXT, mime::EVENT_STREAM)),
		) {
			return Err(tg::error!(?content_type, "invalid content type"));
		}
		let stream = response
			.sse()
			.map_err(|source| tg::error!(!source, "failed to read an event"))
			.and_then(|event| {
				future::ready(
					if event.event.as_deref().is_some_and(|event| event == "error") {
						match event.try_into() {
							Ok(error) | Err(error) => Err(error),
						}
					} else {
						event.try_into()
					},
				)
			});
		Ok(stream)
	}
}

File: packages/clients/js/src/client/checkout.ts
import * as tg from "../index.ts";
import { Body, Request } from "../http.ts";
import type { Client } from "../client.ts";

export namespace Checkout {
	export type Arg = {
		artifact: tg.Artifact.Id;
		dependencies?: boolean;
		extension?: string | null;
		force?: boolean;
		lock?: "auto" | "attr" | "file" | null;
		path?: string | null;
	};

	export namespace Arg {
		export let toJson = (arg: tg.Checkout.Arg): unknown => {
			let output: { [key: string]: unknown } = {
				artifact: arg.artifact,
			};
			if (arg.dependencies !== undefined && !arg.dependencies) {
				output.dependencies = arg.dependencies;
			}
			if (arg.extension !== undefined) {
				output.extension = arg.extension;
			}
			if (arg.force) {
				output.force = arg.force;
			}
			if (arg.lock !== undefined) {
				if (arg.lock === null) {
					output.lock = null;
				} else if (arg.lock !== "auto") {
					output.lock = arg.lock;
				}
			}
			if (arg.path !== undefined) {
				output.path = arg.path;
			}
			return output;
		};
	}

	export type Output = {
		path: string;
	};
}

export async function checkout(
	client: Client,
	arg: tg.Checkout.Arg,
): Promise<AsyncIterableIterator<tg.Progress.Event<tg.Checkout.Output>>> {
	let method = "POST";
	let uri = "/checkout";
	let headers = {
		accept: "text/event-stream",
		"content-type": "application/json",
	};
	let body = Body.json(Checkout.Arg.toJson(arg));
	let request = new Request({
		body,
		method,
		uri,
		headers,
	});
	let response = await client.sendWithRetry(request);
	if (response.status < 200 || response.status >= 300) {
		throw tg.Error.fromData(await response.json<tg.Error.Data>());
	}
	return tg.Progress.decode<tg.Checkout.Output>(response);
}
